### PR TITLE
Test WASM bindings in a Node.js environment

### PR DESCRIPTION
**Pull request by albrow**
This is a WIP PR to add some minimal tests for the JavaScript/WASM bindings. The goal for this PR is to test basic functionality in a Node.js environment. This PR will not include in-browser testing.

### CODE REVIEW

**Comment by albrow** [comment]
Update: You can now run the tests for the WASM bindings with `GOOS=js GOARCH=wasm go test -exec="./go_js_wasm_exec" -v .`

The latest commit enables a shared test (which should work in both Go and the WASM bindings) for `PeerConnection.SetConfiguration`. However, right now this test fails in WASM environments due to failing to convert the error to the expected `pions/webrtc` type.

We need to load in the [`wrtc` package](https://www.npmjs.com/package/wrtc) in order for the WASM bindings to work in Node.js. Doing this required adding some new files: __package.json__, __node_shim.js__, __go_js_wasm_exec__, etc. For now I just threw them into the top-level directory but really we should find a more appropriate place for them.

**Comment by albrow** [comment]
Latest result from Travis:

```
 GOOS=js GOARCH=wasm go test -exec="./go_js_wasm_exec" -v .
panic: newosproc: not implemented
fatal error: panic on system stack
runtime stack:
runtime.throw(0xbe730, 0x15)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/panic.go:617 +0x7
panic(0x45d40, 0x113f60)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/panic.go:460 +0x72
runtime.newosproc(...)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/os_js.go:116
runtime.newm1(0xc044000)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:1861 +0x19
runtime.newm(0xd0490, 0xc02e500)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:1840 +0xc
runtime.startm(0x0, 0x100000001)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:1977 +0x1c
runtime.wakep()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:2058 +0x4
runtime.newproc1(0xd03f8, 0xc03e778, 0x8, 0xc000180, 0x10ab0002)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:3351 +0x48
runtime.newproc.func1()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:3244 +0x2
runtime.systemstack(0x4b97f0)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/asm_wasm.s:171 +0x2
runtime.mstart()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:1153
goroutine 1 [running]:
runtime.systemstack_switch()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/asm_wasm.s:182 fp=0xc03e728 sp=0xc03e720 pc=0x137e0000
runtime.newproc(0x8, 0xd03f8)
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:3243 +0x3 fp=0xc03e768 sp=0xc03e728 pc=0x12400003
runtime.init.0()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/lock_js.go:142 +0x2 fp=0xc03e790 sp=0xc03e768 pc=0x10ab0002
runtime.init()
	<autogenerated>:1 +0x7 fp=0xc03e798 sp=0xc03e790 pc=0x13770007
runtime.main()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/proc.go:147 +0x11 fp=0xc03e7e0 sp=0xc03e798 pc=0x11f30011
runtime.goexit()
	/home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/asm_wasm.s:422 +0x1 fp=0xc03e7e8 sp=0xc03e7e0 pc=0x13a30001
FAIL	github.com/pions/webrtc	6.977s
The command "GOOS=js GOARCH=wasm go test -exec="./go_js_wasm_exec" -v ." exited with 1.
```

This looks pretty gnarly and is not what I see when I run the same command locally. Planning to investigate more on Monday. For context, here's what I see when I run `GOOS=js GOARCH=wasm go test -exec="./go_js_wasm_exec" -v .`:

```
=== RUN   TestNewBundlePolicy
--- PASS: TestNewBundlePolicy (0.00s)
=== RUN   TestBundlePolicy_String
--- PASS: TestBundlePolicy_String (0.00s)
=== RUN   TestConfiguration_getICEServers
=== RUN   TestConfiguration_getICEServers/Success
=== RUN   TestConfiguration_getICEServers/Failure
--- PASS: TestConfiguration_getICEServers (0.00s)
    --- PASS: TestConfiguration_getICEServers/Success (0.00s)
    --- PASS: TestConfiguration_getICEServers/Failure (0.00s)
=== RUN   TestNewDataChannelState
--- PASS: TestNewDataChannelState (0.00s)
=== RUN   TestDataChannelState_String
--- PASS: TestDataChannelState_String (0.00s)
=== RUN   TestDTLSRole_String
--- PASS: TestDTLSRole_String (0.00s)
=== RUN   TestNewDTLSTransportState
--- PASS: TestNewDTLSTransportState (0.00s)
=== RUN   TestDTLSTransportState_String
--- PASS: TestDTLSTransportState_String (0.00s)
=== RUN   TestICECandidate_Convert
--- PASS: TestICECandidate_Convert (0.00s)
=== RUN   TestConvertTypeFromICE
=== RUN   TestConvertTypeFromICE/host
=== RUN   TestConvertTypeFromICE/srflx
=== RUN   TestConvertTypeFromICE/prflx
--- PASS: TestConvertTypeFromICE (0.00s)
    --- PASS: TestConvertTypeFromICE/host (0.00s)
    --- PASS: TestConvertTypeFromICE/srflx (0.00s)
    --- PASS: TestConvertTypeFromICE/prflx (0.00s)
=== RUN   TestICECandidateInit_Serialization
--- PASS: TestICECandidateInit_Serialization (0.00s)
=== RUN   TestICECandidateType
--- PASS: TestICECandidateType (0.00s)
=== RUN   TestICECandidateType_String
--- PASS: TestICECandidateType_String (0.00s)
=== RUN   TestICEComponent
--- PASS: TestICEComponent (0.00s)
=== RUN   TestICEComponent_String
--- PASS: TestICEComponent_String (0.00s)
=== RUN   TestNewICEConnectionState
--- PASS: TestNewICEConnectionState (0.00s)
=== RUN   TestICEConnectionState_String
--- PASS: TestICEConnectionState_String (0.00s)
=== RUN   TestNewICECredentialType
--- PASS: TestNewICECredentialType (0.00s)
=== RUN   TestICECredentialType_String
--- PASS: TestICECredentialType_String (0.00s)
=== RUN   TestICEGathererState_String
--- PASS: TestICEGathererState_String (0.00s)
=== RUN   TestNewICEGatheringState
--- PASS: TestNewICEGatheringState (0.00s)
=== RUN   TestICEGatheringState_String
--- PASS: TestICEGatheringState_String (0.00s)
=== RUN   TestNewICEProtocol
--- PASS: TestNewICEProtocol (0.00s)
=== RUN   TestICEProtocol_String
--- PASS: TestICEProtocol_String (0.00s)
=== RUN   TestNewICERole
--- PASS: TestNewICERole (0.00s)
=== RUN   TestICERole_String
--- PASS: TestICERole_String (0.00s)
=== RUN   TestICEServer_validate
=== RUN   TestICEServer_validate/Success
=== RUN   TestICEServer_validate/Failure
--- PASS: TestICEServer_validate (0.00s)
    --- PASS: TestICEServer_validate/Success (0.00s)
    --- PASS: TestICEServer_validate/Failure (0.00s)
=== RUN   TestNewICETransportPolicy
--- PASS: TestNewICETransportPolicy (0.00s)
=== RUN   TestICETransportPolicy_String
--- PASS: TestICETransportPolicy_String (0.00s)
=== RUN   TestICETransportState_String
--- PASS: TestICETransportState_String (0.00s)
=== RUN   TestICETransportState_Convert
--- PASS: TestICETransportState_Convert (0.00s)
=== RUN   TestCodecRegistration
--- PASS: TestCodecRegistration (0.00s)
=== RUN   TestPeerConnection_SetConfiguration
--- FAIL: TestPeerConnection_SetConfiguration (0.01s)
    peerconnection_test.go:150: SetConfiguration "closed connection": err = JavaScript error: InvalidStateError: RTCPeerConnection is closed, want InvalidStateError: connection closed
    peerconnection_test.go:150: SetConfiguration "update PeerIdentity": err = <nil>, want InvalidModificationError: peerIdentity cannot be modified
    peerconnection_test.go:150: SetConfiguration "update BundlePolicy": err = JavaScript error: InvalidModificationError: , want InvalidModificationError: bundle policy cannot be modified
    peerconnection_test.go:150: SetConfiguration "update RTCPMuxPolicy": err = JavaScript error: InvalidModificationError: , want InvalidModificationError: rtcp mux policy cannot be modified
    peerconnection_test.go:150: SetConfiguration "update ICECandidatePoolSize": err = JavaScript error: InvalidModificationError: , want InvalidModificationError: ice candidate pool size cannot be modified
    peerconnection_test.go:150: SetConfiguration "update ICEServers, no TURN credentials": err = JavaScript error: InvalidAccessError: , want InvalidAccessError: turn server credentials required
=== RUN   TestNewPeerConnectionState
--- PASS: TestNewPeerConnectionState (0.00s)
=== RUN   TestPeerConnectionState_String
--- PASS: TestPeerConnectionState_String (0.00s)
=== RUN   TestNewPriorityType
--- PASS: TestNewPriorityType (0.00s)
=== RUN   TestPriorityType_String
--- PASS: TestPriorityType_String (0.00s)
=== RUN   TestQUICRole_String
--- PASS: TestQUICRole_String (0.00s)
=== RUN   TestNewRTCPMuxPolicy
--- PASS: TestNewRTCPMuxPolicy (0.00s)
=== RUN   TestRTCPMuxPolicy_String
--- PASS: TestRTCPMuxPolicy_String (0.00s)
=== RUN   TestNewRTPTransceiverDirection
--- PASS: TestNewRTPTransceiverDirection (0.00s)
=== RUN   TestRTPTransceiverDirection_String
--- PASS: TestRTPTransceiverDirection_String (0.00s)
=== RUN   TestNewSCTPTransportState
--- PASS: TestNewSCTPTransportState (0.00s)
=== RUN   TestSCTPTransportState_String
--- PASS: TestSCTPTransportState_String (0.00s)
=== RUN   TestNewSDPType
--- PASS: TestNewSDPType (0.00s)
=== RUN   TestSDPType_String
--- PASS: TestSDPType_String (0.00s)
=== RUN   TestSessionDescription_JSON
--- PASS: TestSessionDescription_JSON (0.00s)
=== RUN   TestSetEphemeralUDPPortRange
--- PASS: TestSetEphemeralUDPPortRange (0.00s)
=== RUN   TestSetConnectionTimeout
--- PASS: TestSetConnectionTimeout (0.00s)
=== RUN   TestDetachDataChannels
--- PASS: TestDetachDataChannels (0.00s)
=== RUN   TestNewSignalingState
--- PASS: TestNewSignalingState (0.00s)
=== RUN   TestSignalingState_String
--- PASS: TestSignalingState_String (0.00s)
=== RUN   TestSignalingState_Transitions
--- PASS: TestSignalingState_Transitions (0.00s)
FAIL
FAIL	github.com/pions/webrtc	1.677s
```

File: peerconnection_go_test.go [hunk 1]
```diff
@@ -279,44 +224,6 @@ func TestPeerConnection_GetConfiguration(t *testing.T) {
 // 	assert.Nil(t, err)
 // }
 
-// TODO Fix this test
-const minimalOffer = `v=0
```
Comment (albrow) [inline peerconnection_go_test.go]: The JavaScript WebRTC API doesn't consider this SDP valid and was returning the following error:

```
InvalidAccessError: Failed to set remote offer sdp: The m= section:video is invalid. RTCP-MUX is not enabled when it is required.
```

I updated the SDP in order to fix the issue.

Comment (albrow) [reply peerconnection_go_test.go]: Is `TODO Fix this test` still relevant? Is there something left to do here?

**Comment by albrow** [comment]
TravisCI failure looks like it has something to do with this issue: https://travis-ci.community/t/goos-js-goarch-wasm-go-run-fails-panic-newosproc-not-implemented/1651.

**Comment by backkem** [comment]
Hmm, weird. The same approach seemed to fix the problems for gopherwasm: https://github.com/gopherjs/gopherwasm/commit/d9eb0f4c0e2e08c9c9d573df57020f39f008545f

**Comment by albrow** [comment]
@backkem good idea. I removed the default installation step for Go and it seems to work now 🎉

**Comment by albrow** [comment]
@backkem I still want to add tests for `DataChannel`, but at this point I'm satisfied with the approach, file structure, CI configuration, etc. If you want, feel free to go ahead and review those aspects. Otherwise I'll ping you again once the `DataChannel` tests are in place.

**Comment by albrow** [comment]
@backkem I added the remaining tests for `DataChannel`. This PR is ready for review.

**Comment by albrow** [comment]
> Maybe we can document/ticket out the required changes to unify to JS and Go tests again were possible.

@backkem good idea. I created #519 with all the API differences I know about right now. We can add more to that issue as we find more differences.

**Pull request merged**